pkg/api/servicecidr/servicecidr.go

```diff
@@ -0,0 +1,118 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package servicecidr
+
+import (
+	"fmt"
+	"net"
+	"net/netip"
+
+	networkingv1alpha1 "k8s.io/api/networking/v1alpha1"
+	"k8s.io/apimachinery/pkg/labels"
+	networkinglisters "k8s.io/client-go/listers/networking/v1alpha1"
+)
+
+// ContainsIP return the list of ServiceCIDR that contains the IP address passed as argument
+func ContainsIP(serviceCIDRLister networkinglisters.ServiceCIDRLister, ip net.IP) []*networkingv1alpha1.ServiceCIDR {
+	address := IPToAddr(ip)
+	return ContainsAddress(serviceCIDRLister, address)
+}
+
+// ContainsAddress return the list of ServiceCIDR that contains the address passed as argument
+func ContainsAddress(serviceCIDRLister networkinglisters.ServiceCIDRLister, address netip.Addr) []*networkingv1alpha1.ServiceCIDR {
+	result := []*networkingv1alpha1.ServiceCIDR{}
+	serviceCIDRList, err := serviceCIDRLister.List(labels.Everything())
+	if err != nil {
+		return result
+	}
+
+	for _, serviceCIDR := range serviceCIDRList {
+		for _, cidr := range serviceCIDR.Spec.CIDRs {
+			if prefix, err := netip.ParsePrefix(cidr); err == nil { // it can not fail since is already validated
+				if prefixContainsIP(prefix, address) {
+					result = append(result, serviceCIDR)
+				}
+			}
+		}
+	}
+	return result
+}
+
+// prefixContainsIP returns true if the given IP is contained with the prefix,
+// is not the network address and also, if IPv4, is not the broadcast address.
+// This is required (rather than just `prefix.Contains(ip)`) because a ServiceCIDR
+// covering prefix will not allocate those IPs, so a service with one of those IPs
+// can't belong to that ServiceCIDR.
+func prefixContainsIP(prefix netip.Prefix, ip netip.Addr) bool {
+	// if the IP is the network address is not contained
```

Comment:
we already discussed this I can't' remember where and when, if my mind does not fail, we concluded:
- you are technically right
- the rest of the world may not be technically right and may do assumptions on these ips and not work correctly

it will be change for sure, will it be breaking? if some service implementation makes the same assumptions, I think IPVS may , we may end being technically correct but not implementable depending on project or technology.

I don't think at this point this is a problem or worth to change

Reply:
LOL, I do not recall that conversation, but it's plausible.  I guess we set precedent by apiserver being .1 instead of .0.

Can we document this in API comments for `ServiceCIDRSpec.CIDRs`.  E.g. "The first and last IP of each CIDR are reserved, and will not be allocated"

Reply:
for IPv6 the last is not reserved 🥇

Reply:
added an additional comment documenting the current details about allocatable IPs